cargo-miri/bin.rs

```diff
@@ -729,6 +786,57 @@ fn phase_cargo_runner(binary: &Path, binary_args: env::Args) {
     if verbose {
         eprintln!("[cargo-miri runner] {:?}", cmd);
     }
+
+    if std::env::var_os("MIRI_CALLED_FROM_RUSTDOC").is_some() {
+        exec_with_pipe(cmd, &info.stdin)
+    } else {
+        exec(cmd)
+    }
+}
+
+fn phase_cargo_rustdoc(fst_arg: &str, mut args: env::Args) {
+    let verbose = std::env::var_os("MIRI_VERBOSE").is_some();
+
+    // phase_cargo_miri sets the RUSTDOC env var to ourselves, so we can't use that here;
+    // just default to a straight-forward invocation for now:
+    let mut cmd = Command::new(OsString::from("rustdoc"));
+
+    let extern_flag = "--extern";
+    assert!(fst_arg != extern_flag);
+    cmd.arg(fst_arg);
+
+    // Patch --extern arguments to use *.rmeta files, since phase_cargo_rustc only creates stub *.rlib files.
+    while let Some(arg) = args.next() {
+        if arg == extern_flag {
+            cmd.arg(extern_flag); // always forward flag, but adjust filename
+            // `--extern` is always passed as a separate argument by cargo.
+            let next_arg = args.next().expect("`--extern` should be followed by a filename");
+            if let Some(next_lib) = next_arg.strip_suffix(".rlib") {
+                // If this is an rlib, make it an rmeta.
+                cmd.arg(format!("{}.rmeta", next_lib));
+            } else {
+                // Some other extern file (e.g., a `.so`). Forward unchanged.
+                cmd.arg(next_arg);
+            }
+        } else {
+            cmd.arg(arg);
+        }
+    }
+
+    cmd.arg("-Z").arg("unstable-options");
```

Comment:
Presumably this is needed for `--test-builder`/`--runtool`? Please add a comment along those lines.